doc/15-troubleshooting.md

```diff
@@ -230,6 +232,91 @@ icinga2 7764 nagios  mem    REG                8,5   744776 2631636 /usr/lib/x86
 #
 ```
 
+### Optimise apply rules and group assign conditions
+
+#### Remove actually unused apply rules
+
+If `icinga2 daemon -C` warns you like shown below and the respective apply rule
+should indeed create no objects, consider removing it or commenting it out.
+
+```
+[...] warning/ApplyRule: Apply rule '...' (...) for type '...' does not match anywhere!
+```
+
+At least consider changing the `assign where` filter from, say
+`assign where my_complex_filter()`, to `assign where false && my_complex_filter()`.
+`&&` will evaluate `false` first and in an instant.
+If it's not true, `&&` will "return" `false` and skip `my_complex_filter()`.
+
+Same for `assign where` conditions in groups. In this case removing just
+the `assign where` line(s) is enough if you'd like to keep the group itself.
+
+#### Avoid creating single objects via apply rules
+
+If possible, replace constructs like the immediately following with the below one.
+
+```
+apply Service "syspatch" {
+	// ...
+	assign where host.name == "firewall"
+}
+```
+
+```
+object Service "syspatch" {
+	// ...
+	host_name = "firewall"
+}
+```
+
+This way Icinga won't have to evaluate `host.name == "firewall"`
+for all of your hosts.
+
+At least upgrade to Icinga v2.13.6+ which optimizes apply rules as shown above.
+(In this case such `assign where` filters even save time compared to e.g.
+`assign where host.vars.os == "OpenBSD"` if such hosts are a minority.
+Useful if the above recommendation is not an option.)
+
+Group membership assign conditions aren't optimized by v2.13.6.
+But they can just be replaced with constructs like this one:
+
+```
+object Host "firewall" {
+	// ...
+	groups = [ "OpenBSD hosts" ]
+}
+```
+
+#### Reduce `assign where` filter complexity
+
+If neither removals, nor no-op conditions (`assign where false && ...`),
+nor O(1) conditions (`assign where host.name == "firewall"`)
+are an option, keep the filter as simple as possible.
```

Comment:
Just skip the O notation, it's not very useful here. On the level of evaluating a single expression on a single object, we are fighting the constant factors, something that O notation completely ignores. (On the level of evaluating apply rules as a whole it matters, this is O(#objects * #rules), so that's why things get slow if both grow.)

As soon as you do any operation that somewhere does a lookup in a `std::map`, you're technically in O(log n), but this is not what's causing things to be slow here.